nixos/prometheus: add `ec2_sd_configs` section to `scrape_configs`

thefloweringash opened the pull request:
###### Motivation for this change

Prometheus's configuration in Nix is exhaustively specified. To use any of the various service discovery mechanisms they must be defined in the module. This adds definitions for the EC2 service discovery according to the [Prometheus Documentation for version 1.8][], matching the default value of `config.services.prometheus.package`.

Applies cleanly to 18.09 where I tested it, and is purely additive. If merged, I'd like it to be backported as well.

In the interests of not having to closely track upstream changes, would it be preferable to have an `extraConfigs` inside `scrapeConfigs`, that is merged as pure data? For example, if a user wants to use prometheus 2.6, it has a new option in the EC2 service discovery section called `filters`. As is, this is impossible(?) without updating this module again, but with an `extraConfigs` there's an escape.

[Prometheus Documentation for version 1.8]: https://prometheus.io/docs/prometheus/1.8/configuration/configuration/#ec2_sd_config

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Assured whether relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

thefloweringash committed:
nixos/prometheus: add `ec2_sd_configs` section to `scrape_configs`

basvandijk reviewed (changes requested):
@thefloweringash note I just merged PR https://github.com/NixOS/nixpkgs/pull/58255 which adds the `prometheus2` service. 

I see that the `ec2_sd_config` options are different between [1.8](https://prometheus.io/docs/prometheus/1.8/configuration/configuration/#ec2_sd_config) and [2.8](https://prometheus.io/docs/prometheus/2.8/configuration/configuration/#ec2_sd_config). So I think it makes sense starting to split the options for these version. Like having both `promTypes.ec2_sd_config_v1` and `promTypes.ec2_sd_config_v2`. 

I also agree it would be nice to have a  `extraConfigs` in `scrapeConfigs` like you suggest.

basvandijk committed:
nixos/prometheus: add new ec2_sd_config options for prometheus2

basvandijk committed:
nixos/prometheus: filter out empty srcape_configs attributes

This results in a smaller prometheus.yml config file.

It also allows us to use the same options for both prometheus-1 and
prometheus-2 since the new options for prometheus-2 default to null
and will be filtered out if they are not set.
